[inductor] Store real inputs to be used for cpp wrapper codegen

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103289

Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @aakhundov

desertfire committed:
[inductor] Store real inputs to be used for cpp wrapper codegen

Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

[ghstack-poisoned]

eellison reviewed (commented):
some overlap with https://github.com/pytorch/pytorch/pull/100652

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -736,12 +735,14 @@ def compile_fx(
 
     if config.cpp_wrapper:
-        # CudaWrapperCodeGen relies on kernel name to find the autotuned cubin file
+        V.set_real_inputs(example_inputs_)
```
Perhaps use a `with....` syntax so this gets cleared when compile is done.

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Store real inputs to be used for cpp wrapper codegen"


Summary: defaked args (zeros) may cause device-side access assertion, so
record the orginal real tensor inputs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 aakhundov

[ghstack-poisoned]

eellison commented on test/inductor/test_torchinductor.py:
```diff
@@ -4405,6 +4405,24 @@ def fn(ind, x, src):
+            ne = torch.ops.aten.ne.Scalar(x, 0)
+            sum = torch.ops.aten.sum.dim_IntList(ne, [-1])
+            sub = torch.ops.aten.sub.Tensor(sum, 1)
+            iota = torch.ops.prims.iota.default(
```
also add a test where you index with a param ?

eellison commented on torch/_inductor/compile_fx.py:
```diff
@@ -736,14 +737,15 @@ def compile_fx(
+                "triton.unique_kernel_names": True,
             }
-        ):
+        ), V.set_real_inputs(example_inputs_):
```
one thing that's potentially confusing is `example_inputs_` here is only going to not going to include non-Params.. will that work ?

desertfire replied:
params come from `torch._guards.TracingContext.get().params_flat` which was added by your PR.

eellison replied:
Didn't you say `found the pure AOTInductor test (not cpp_wrapper tests) failed because inductor is called directly without going through dynamo and thus torch._guards.TracingContext`

so that won't always be set right ?

desertfire commented:
@pytorchbot rebase -b main

desertfire commented:
@pytorchbot merge

desertfire commented:
@pytorchbot merge -f "ignore trunk failures"
